### PR TITLE
Fix H2 Server with SNI bug

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
H2 server with SNI enabled current does not work. Netty's SniHandler
does not queue writes in the same way that the SslHandler does. The h2
handler is added and writes a settings frame immediately on the server,
which is written to the socket before the handshake is completed. This
results in a handshake failure.

Modifications:
- Defer adding h2 protocol handlers to the pipeline until after the SNI
  selection is made.

Result:
H2 Server with SNI now works as expected.

### CODE REVIEW

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SniCompleteChannelSingle.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.ssl.SniCompletionEvent;
+
+import java.nio.channels.ClosedChannelException;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.transport.netty.internal.ChannelCloseUtils.assignConnectionError;
+
+final class SniCompleteChannelSingle extends ChannelInitSingle<SniCompletionEvent> {
+    private final boolean forceRead;
```
Comment (chemicL) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SniCompleteChannelSingle.java:33]: This flag is always true in the current usage. Is it necessary? If so, a test demonstrating the need for a false variant would be useful.

Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SniCompleteChannelSingle.java:33]: good point, I will remove this flag. This was taken from `AlpnChannelSingle` which is used on the client (doesn't need force read) and server (needs force read), however in this case we only need to cover the server use case.

**Comment by Scottmitch** [comment]
test failure attributed to https://github.com/apple/servicetalk/issues/1663

**Review by idelpivnitskiy — APPROVED** [review]
Approach LGTM!
Only minor suggestions and a few questions:

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SniCompleteChannelSingle.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                subscriber = null;
+                subscriberCopy.onSuccess((SniCompletionEvent) evt);
+            }
+            ctx.fireUserEventTriggered(evt);
+        }
+
+        @Override
+        public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+            if (subscriber != null) {
+                Exception e = new ClosedChannelException();
```
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SniCompleteChannelSingle.java:73]: sure

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SniCompleteChannelSingle.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                assignConnectionError(ctx.channel(), e);
+                final SingleSource.Subscriber<? super SniCompletionEvent> subscriberCopy = subscriber;
+                subscriber = null;
+                subscriberCopy.onError(e);
+            } else {
+                // Propagate exception in the pipeline if subscriber is already complete
```
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SniCompleteChannelSingle.java:79]: in practice I don't think it matters bcz the connection will be torn down and the pipeline will be cleaned up anyway.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DeferredServerChannelBinder.java [hunk 4]
```diff
@@ -97,4 +102,30 @@ private AlpnServerContext() {
...
+
+            if (config.h2Config() != null) {
+                return H2ServerParentConnectionContext.initChannel(listenAddress, channel, httpExecutionContext, config,
+                        NoopChannelInitializer.INSTANCE, service, drainRequestPayloadBody, observer);
+            } else if (config.h1Config() != null) {
+                return NettyHttpServer.initChannel(channel, httpExecutionContext, config,
+                        NoopChannelInitializer.INSTANCE, service, drainRequestPayloadBody, observer);
+            }
+            return failed(new IllegalStateException(
+                    "SSL handshake completed, but no protocols to initialize. Consider using ALPN."));
```
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DeferredServerChannelBinder.java:128]: ALPN would make the protocol exchange / support explicit (requiring configuring up front). Here we don't know which protocol is "preferred" and give precedence to h2.

**Comment by Scottmitch** [comment]
build failure attributed to https://github.com/apple/servicetalk/issues/1507

**Comment by Scottmitch** [comment]
0.41 (05a1cc5ca)

**Pull request merged**